src/services/combinePath.ts

```diff
@@ -1,26 +1,31 @@
 import { RemoveLeadingQuestionMarkFromKeys } from '@/types/params'
-import { Path, PathParamsWithParamNameExtracted, ToPath } from '@/types/path'
 import { checkDuplicateParams } from '@/utilities/checkDuplicateKeys'
+import { ParamsWithParamNameExtracted, ToWithParams, withParams, WithParams } from '@/services/withParams'
+import { StringHasValue } from '@/utilities/guards'
+
+type CombinePathString<TParent extends string | undefined, TChild extends string | undefined> =
+  StringHasValue<TParent> extends true
+    ? StringHasValue<TChild> extends true
+      ? `${TParent}${TChild}`
+      : TParent
+    : TChild
 
 export type CombinePath<
-  TParent extends Path,
-  TChild extends Path
-> = ToPath<TParent> extends { value: infer TParentPath extends string, params: infer TParentParams extends Record<string, unknown> }
-  ? ToPath<TChild> extends { value: infer TChildPath extends string, params: infer TChildParams extends Record<string, unknown> }
-    ? RemoveLeadingQuestionMarkFromKeys<TParentParams> & RemoveLeadingQuestionMarkFromKeys<TChildParams> extends PathParamsWithParamNameExtracted<`${TParentPath}${TChildPath}`>
-      ? Path<`${TParentPath}${TChildPath}`, RemoveLeadingQuestionMarkFromKeys<TParentParams> & RemoveLeadingQuestionMarkFromKeys<TChildParams>>
-      : Path<'', {}>
-    : Path<'', {}>
-  : Path<'', {}>
+  TParent extends WithParams,
+  TChild extends WithParams
+> = ToWithParams<TParent> extends { value: infer TParentPath extends string | undefined, params: infer TParentParams extends Record<string, unknown> }
+  ? ToWithParams<TChild> extends { value: infer TChildPath extends string | undefined, params: infer TChildParams extends Record<string, unknown> }
+    ? RemoveLeadingQuestionMarkFromKeys<TParentParams> & RemoveLeadingQuestionMarkFromKeys<TChildParams> extends ParamsWithParamNameExtracted<CombinePathString<TParentPath, TChildPath>>
```

Comment:
at some point it would be nice to change the way we're tracking params as optional. We're keeping the `?` in the param name but then we do a bunch of stuff to both "keep it" and "remove it" depending on the types we want. 

I think we should maybe update `WithParams` to return something like
```ts
{
  value: string,
  params: Record<string, param>,
  optional: string[]
}
```
Food for thought for future us

Reply:
I'll add this to my personal list and explore further, would love to get rid of leading "?" in param names